### PR TITLE
[WIP] adapter: reconcile storage commands

### LINKED ISSUE #12096: adapter: storage command reconciliation
### Feature request

Follow up of #10549 (much of the description is copied from there and just changed to be applicable to the storage node instead of compute)

Start a design / implementation for reconciliation of storage commands. We expect both the storage controller and storage instance to go down and up. When they do, they need to re-establish the correct set-up. The controller is probably the source of truth for what sources are valid and what their schemas are, but the storage instance is the source of truth for the contents of the sources.

Currently the controller will always attempt to rebuild every source from scratch and populate catalog tables on startup, which causes the storage instance to fail because those sources already exist.

### PR DESCRIPTION
**Pull request by jkosh44**
When the ADAPTER starts up, it attempts to re-create all sources and
populate all built-in catalog tables. If STORAGE has also just started
up, then this works fine. However if STORAGE was already running and
installed all the sources from a previous ADAPTER, STORAGE will crash
because it already has all of the sources created.

This commit attempts to resolve the above issue by adding a
reconciliation layer in front of the storage instance. The
reconciliation layer will absorb previously seen commands.

The ADAPTER will always clear the contents of all built-in catalog
tables and re-populate them from scratch every time it restarts.

This commit does not attempt to reconcile the RenderSources command
because that command is only sent between STORAGE and COMPUTE. So it is
saved for a later commit.

Resolves #12096

### Motivation
This PR adds a known-desirable feature.

### Testing

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - There are no user-facing behavior changes.

### CODE REVIEW

**Comment by jkosh44** [comment]
All I've done for this PR is to update @antiguru's old code to resolve all merge conflicts and replace the `Insert` command with `Append`. Also I've added the plumbing to pass the `--reconcile` flag to `storaged` and `computed`.

It actually seems to work pretty well with at least basic usages. I can create a table, insert data, create a materialized view over the table, restart `materialized` and nothing blows up.

I am having a little trouble writing automated tests, there is one "gross hack" left in the command reconciliation logic, and there can probably be some code duplication removal across storage and compute reconciliation. Other than that I don't have a good understanding of what's holding storage command reconciliation back, so I'm curious to hear what else are major blockers for this.

File: src/dataflow-types/src/reconciliation/command.rs [hunk 1]
```diff
@@ -35,9 +40,45 @@ use tracing::warn;
...
+use crate::client::{
+    ComputeClient, ComputeCommand, ComputeResponse, GenericClient, StorageClient, StorageCommand,
+    StorageResponse, TimestampBindingFeedback,
+};
+use crate::sources::SourceConnector;
 use crate::{DataflowDescription, Plan};
 
+/// Generic trait to reconcile commands targeted at a COMPUTE or STORAGE instance.
+#[async_trait]
+pub trait CommandReconcile<C, R>: fmt::Debug + Send
```
Comment (antiguru) [inline src/dataflow-types/src/reconciliation/command.rs:52]: I like the intention, but I'm not sure this is in the spirit we would like to evolve storage and compute. In principle, the storage command reconciliation should live in the storage crate, as it doesn't exist in the controller. Extracting a common trait makes this more complicated. Maybe @petrosagg has some opinion on this.

**Review by antiguru — COMMENTED** [review]
I added a comment, but would like to do a full review next week once I'm back.

File: src/dataflow-types/src/reconciliation/command.rs (hunk 1 above)
Comment (antiguru) [reply src/dataflow-types/src/reconciliation/command.rs:52]: Related: https://github.com/MaterializeInc/materialize/issues/12211

Comment (jkosh44) [reply src/dataflow-types/src/reconciliation/command.rs]: Yeah that makes sense, I reverted the commit.

**Pull request closed without merging**